### PR TITLE
Implemented Threads

### PR DESCRIPTION
**Pull request by DWaffles**
# Summary
This is WIP PR for implementation of threads. Discord API documentation is at times wrong or currently in the middle of being PR'd to fix, but I will continue working on this. PR'ing this so I can get feedback. This is still somewhat of a rough WIP so there are a few loose comments that will be removed. This brings the lib up API v9 to note.

I've been using [these commands](https://gist.github.com/DWaffles/51d04454ac881751a97d198e2e8cccae) to test the changes if you wish to see how someone can programmatically use threads. 

# Details
The implementation of this method is that DiscordThreadChannel extends DiscordChannel, allowing easy integration with the rest of the library with small changes. To use thread methods, a DiscordChannel just needs to be to casted to a DiscordThreadChannel. For commands next, the DiscordChannel overload works properly with this. 

Thread cache has been implemented and it is the goal that all active threads will be cached. I am still trying to test edge cases. How this is implemented is that DiscordGuild has a `_threads` property. I make the assumption that all active threads will be cached, as this is how Discord seems to do things.

I also make the assumption that guild features are always properly cached which they should be, this is to check if people can create private threads or use the three and seven day archive options.

# Changes

## Added Events

- [x] ThreadCreated
- [x] ThreadUpdated
- [x] ThreadDeleted
- [x] ThreadListSynced
- [x] ThreadMemberUpdated
- [x] ThreadMembersUpdated

## Endpoints
- [x] Start Thread with Message
- [x] Start Thread without Message
- [x] Join Thread
- [x] Leave Thread
- [x] Add Thread Member
- [x] Remove Thread Member
- [x] List Thread Members
- [x] List Active Threads
- [x] List Public Archived Threads
- [x] List Private Archived Threads
- [x] List Joined Private Archived Threads

### DiscordGuild
* Threads property & _threads
* ListActiveThreadsAsync()

### DiscordChannel
* IsThread property
* Threads property - using DiscordGuild's `_threads`
* TriggerTypingAsync will not throw within thread channels
* CreateThreadAsync() plus overloads

### DiscordMessage
* Channel property modified to work with threads
* CreateThreadAsync()

### Other
* ChannelType enum has new thread types
* GetChannelAsync() will return a thread object as a DiscordChannel if the channel is type thread

### DiscordThreadChannel
* CurrentMember refers to the bot. If the bot has joined the thread, this will have a value
* JoinThreadAsync(), LeaveThreadAsync(), ListJoinedMembersAsync(), AddThreadMemberAsync(), RemoveThreadMemberAsync()

### Other
* DiscordThreadMetadata DiscordThreadMember have base their properties from the API
* AutoArchiveDuration representing... the... auto archive duration of a channel.
* RestThreadCreatePayload

# Still To-Do
- Create flags enum 
- Thread edit capabilities
- Document/add checks on DiscordChannel methods that will fail/throw when it's a threads
- Properly test/document exceptions for new methods
- Audit log stuff for threads
- Do all that fancy operator overloading for created entities

# Notes
* I can no longer test any features locked behind nitro boosts.
* Regarding thread member update, @ajpalkovic: "this event is documented for completeness, but unlikely to be used by most bots. For bots, this event largely is just a signal that you are a member of the thread." So it's here for completeness. I have no idea when it's called anyway. Whenever the bot is added or removed from a thread, thread *members* gets called.
* If anyone can advise on how to better access member objects than the [current way](https://github.com/DSharpPlus/DSharpPlus/pull/1009/files#diff-5c4c3bc5e4d8627351338d1d22cb3901a47a7e63e3ff0acff77b3e59de5e7a36R1964-R1967), let me know.

### CODE REVIEW

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 1]
```diff
@@ -1775,6 +1832,175 @@ internal async Task OnVoiceServerUpdateEventAsync(string endpoint, string token,
...
+        }
+
+        internal async Task OnThreadMemberUpdateEventAsync(DiscordThreadChannelMember member)
+        {
+            member.Discord = this;
+
+            var thread = this.InternalGetCachedThread(member.ThreadId);
+            member._guild_id = thread.Guild.Id;
+            thread.CurrentMember = member;
+            thread.Guild._threads[member.ThreadId] = thread;
```
Comment (VelvetThePanda) [inline DSharpPlus/Clients/DiscordClient.Dispatch.cs:1948]: .AddOrUpdate

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 2]
```diff
@@ -1775,6 +1832,175 @@ internal async Task OnVoiceServerUpdateEventAsync(string endpoint, string token,
...
+            await this._threadListSynced.InvokeAsync(this, new ThreadListSyncEventArgs { Guild = guild, Channels = channels.ToList().AsReadOnly(), Threads = threads, CurrentMembers = members.ToList().AsReadOnly() }).ConfigureAwait(false);
+        }
+
+        internal async Task OnThreadMemberUpdateEventAsync(DiscordThreadChannelMember member)
+        {
+            member.Discord = this;
+
+            var thread = this.InternalGetCachedThread(member.ThreadId);
+            member._guild_id = thread.Guild.Id;
+            thread.CurrentMember = member;
```
Comment (VelvetThePanda) [inline DSharpPlus/Clients/DiscordClient.Dispatch.cs:1947]: Shouldn't this follow the [behavior defined in DiscordGuild.cs](https://github.com/DSharpPlus/DSharpPlus/blob/master/DSharpPlus/Entities/Guild/DiscordGuild.cs#L464)?

**Review by VelvetThePanda — CHANGES_REQUESTED** [review]
There's a lot to be reviewed. I'll review the rest when I have time, but in the meantime cc @Neuheit @Emzi0767

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 3]
```diff
@@ -346,6 +347,39 @@ internal async Task HandleDispatchAsync(GatewayPayload payload)
 
                 #endregion
 
+                #region Thread
+
+                case "thread_create":
+                    trd = dat.ToDiscordObject<DiscordThreadChannel>();
```
[the remark this answers is not in the document]
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:353]: The client is set within OnThreadCreateEventAsync() itself as it's done with channel create/update/delete args. Though I can move it here.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 4]
```diff
@@ -346,6 +347,39 @@ internal async Task HandleDispatchAsync(GatewayPayload payload)
...
+                    await this.OnThreadCreateEventAsync(trd).ConfigureAwait(false);
+                    break;
+
+                case "thread_update":
+                    trd = dat.ToDiscordObject<DiscordThreadChannel>();
```
[the remark this answers is not in the document]
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:358]: Set further, per previous comment.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 5]
```diff
@@ -346,6 +347,39 @@ internal async Task HandleDispatchAsync(GatewayPayload payload)
...
+                    await this.OnThreadUpdateEventAsync(trd).ConfigureAwait(false);
+                    break;
+
+                case "thread_delete":
+                    trd = dat.ToDiscordObject<DiscordThreadChannel>();
```
[the remark this answers is not in the document]
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:363]: Set further, per previous comment.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 6]
```diff
@@ -500,6 +536,11 @@ internal async Task OnReadyEventAsync(ReadyPayload ready, JArray rawGuilds, JArr
                         xo._channel_id = xc.Id;
                     }
                 }
+                foreach (var xc in guild.Threads.Values)
```
[the remark this answers is not in the document]
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:539]: Done, to be committed soon.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (VelvetThePanda) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:353]: I thought these were invoking events, oops. Keep it where it is

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 7]
```diff
@@ -1775,6 +1832,175 @@ internal async Task OnVoiceServerUpdateEventAsync(string endpoint, string token,
 
         #endregion
 
+        #region Thread
+
+        internal async Task OnThreadCreateEventAsync(DiscordThreadChannel thread)
+        {
+            thread.Discord = this;
+            this.InternalGetCachedGuild(thread.GuildId)._threads[thread.Id] = thread;
```
[the remark this answers is not in the document]
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1840]: Done, to be committed.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 8]
```diff
@@ -1775,6 +1832,175 @@ internal async Task OnVoiceServerUpdateEventAsync(string endpoint, string token,
...
+                };
+                guild._threads[thread.Id] = thread;
+            }
+
+            await this._threadUpdated.InvokeAsync(this, updateEvent).ConfigureAwait(false);
+        }
+
+        internal async Task OnThreadDeleteEventAsync(DiscordThreadChannel thread)
+        {
+            if (thread == null)
```
[the remark this answers is not in the document]
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1903]: This is akin to the current behavior of ChannelDeleteEventAsync, I think we'll discuss this in Discord and I'll follow yours and the other maintainers suggestion on this.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 9]
```diff
@@ -1775,6 +1832,175 @@ internal async Task OnVoiceServerUpdateEventAsync(string endpoint, string token,
...
+            foreach(var member in members)
+            {
+                member.Discord = this;
+                member._guild_id = guild.Id;
+                var thread = threads.SingleOrDefault(x => x.Id == member.ThreadId);
+                if (thread != null)
+                    thread.CurrentMember = member;
+            }
+
+            await this._threadListSynced.InvokeAsync(this, new ThreadListSyncEventArgs { Guild = guild, Channels = channels.ToList().AsReadOnly(), Threads = threads, CurrentMembers = members.ToList().AsReadOnly() }).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1938]: They are technically readonly but they implement methods that allow for mutation unless I'm mistaken. Regarding IReadOnlyCollection vs IReadOnlyList I was informed by @Emzi0767 to use IReadOnlyList.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs (hunk 1 above)
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1948]: Done, to be committed.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs (hunk 2 above)
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1947]: Not to my understanding. Current member is a singular member representing whether or not the bot is in the thread. Its this way atm to stay more consistent with Discord's API.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 10]
```diff
@@ -1775,6 +1832,175 @@ internal async Task OnVoiceServerUpdateEventAsync(string endpoint, string token,
...
+                guild._threads[thread.Id] = thread;
+            }
+            foreach(var member in members)
+            {
+                member.Discord = this;
+                member._guild_id = guild.Id;
+                var thread = threads.SingleOrDefault(x => x.Id == member.ThreadId);
+                if (thread != null)
+                    thread.CurrentMember = member;
+            }
```
[the remark this answers is not in the document]
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1936]: Added some spacing, to be committed. Setting the thread current member can be done in the thread loop, it'll be a single line in comparison if that's preferable, though technically it's a bit less efficient.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 11]
```diff
@@ -1775,6 +1832,175 @@ internal async Task OnVoiceServerUpdateEventAsync(string endpoint, string token,
...
+            else
+            {
+                updateEvent = new ThreadUpdateEventArgs
+                {
+                    ThreadAfter = thread,
+                    Guild = thread.Guild,
+                    Parent = thread.Parent
+                };
+                guild._threads[thread.Id] = thread;
+            }
```
[the remark this answers is not in the document]
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1896]: I set the optional because to me it makes sense that this field is more akin to boolean value if the user is in the thread or not. That is of course, me. I'll change it not use optional within my next few commits.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs]: Is this more clear?

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ProfDoof) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1943]: Yeah, when you mark them AsReadOnly() it wraps the list so that you can no longer add or remove things from the list. If you don't do that, then they still expose those methods even if you have to down(?)cast them to do so.

[the remark this answers is not in the document]
Comment (VelvetThePanda) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1943]: TIL it actually [generates a new ReadOnlyCollection](https://source.dot.net/#System.Private.CoreLib/List.cs,247). Interesting

File: DSharpPlus/Clients/DiscordClient.Events.cs [hunk 12]
```diff
@@ -568,6 +568,76 @@ public sealed partial class DiscordClient
...
+        /// </summary>
+        public event AsyncEventHandler<DiscordClient, ThreadDeleteEventArgs> ThreadDeleted
+        {
+            add => this._threadDeleted.Register(value);
+            remove => this._threadDeleted.Unregister(value);
+        }
+        private AsyncEvent<DiscordClient, ThreadDeleteEventArgs> _threadDeleted;
+
+        /// <summary>
+        /// Fired when a thread member is updated.
```
Comment (IDoEverything) [inline DSharpPlus/Clients/DiscordClient.Events.cs:607]: Wrong description

File: DSharpPlus/Clients/DiscordShardedClient.Events.cs [hunk 13]
```diff
@@ -554,6 +554,76 @@ public sealed partial class DiscordShardedClient
...
+        /// </summary>
+        public event AsyncEventHandler<DiscordClient, ThreadDeleteEventArgs> ThreadDeleted
+        {
+            add => this._threadDeleted.Register(value);
+            remove => this._threadDeleted.Unregister(value);
+        }
+        private AsyncEvent<DiscordClient, ThreadDeleteEventArgs> _threadDeleted;
+
+        /// <summary>
+        /// Fired when a thread member is updated.
```
Comment (IDoEverything) [inline DSharpPlus/Clients/DiscordShardedClient.Events.cs:593]: Same here

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 14]
```diff
@@ -179,6 +186,20 @@ public IEnumerable<DiscordChannel> Children
             }
         }
 
+        /// <summary>
+        /// Gets this channel's threads. This applies only to text and news channels.
+        /// </summary>
+        [JsonIgnore]
+        public IEnumerable<DiscordThreadChannel> Threads
```
Comment (IDoEverything) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:193]: This should return `IReadOnlyList` not `IEnumerable`

File: DSharpPlus/Entities/Channel/Thread/DiscordThreadChannel.cs [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+        /// Gets the approximate count of members in a thread, capped to 50.
+        /// </summary>
+        [JsonProperty("member_count", NullValueHandling = NullValueHandling.Ignore)]
+        public int? MemberCount { get; internal set; }
+
+        /// <summary>
+        /// Represents the current member for this thread. This will have a value if the user has joined the thread.
+        /// </summary>
+        [JsonProperty("member", NullValueHandling = NullValueHandling.Ignore)]
+        public Optional<DiscordThreadChannelMember> CurrentMember { get; internal set; }
```
Comment (IDoEverything) [inline DSharpPlus/Entities/Channel/Thread/DiscordThreadChannel.cs:65]: shouldn't this be nullable, not `Optional<>`?

File: DSharpPlus/Entities/Channel/Thread/DiscordThreadChannelMetadata.cs [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    public class DiscordThreadChannelMetadata
+    {
+        /// <summary>
+        /// Gets whether this thread is archived or not.
+        /// </summary>
+        [JsonProperty("archived", NullValueHandling = NullValueHandling.Ignore)]
+        public bool IsArchived { get; internal set; }
+
+        /// <summary>
+        ///	Gets the duration in minutes to automatically archive the thread after recent activity. Can be set to: 60, 1440, 4320, 10080.
```
Comment (IDoEverything) [inline DSharpPlus/Entities/Channel/Thread/DiscordThreadChannelMetadata.cs:47]: Extra space in the beginning

File: DSharpPlus/Enums/Channel/AutoArchiveDuration.cs [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+// SOFTWARE.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace DSharpPlus
+{
+    /// <summary>
+    ///	Represents the duration in minutes to automatically archive a thread after recent activity.
```
Comment (IDoEverything) [inline DSharpPlus/Enums/Channel/AutoArchiveDuration.cs:31]: Spaces

File: DSharpPlus/EventArgs/Channel/Thread/ThreadListSyncEventArgs.cs [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace DSharpPlus.EventArgs
+{
+    /// <summary>
+    /// Represents arguments for <see cref="DiscordClient.ThreadListSynced"/> event.
+    /// </summary>
+    public class ThreadListSyncEventArgs : DiscordEventArgs
+    {
+        /// <summary>
+        /// Gets all thread member objects, indicating which threads the current user has been added to
```
Comment (IDoEverything) [inline DSharpPlus/EventArgs/Channel/Thread/ThreadListSyncEventArgs.cs:35]: Full stop

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 19]
```diff
@@ -842,6 +915,58 @@ public override string ToString()
...
+            else if (message.ChannelId != this.Id)
+                throw new ArgumentException("You must use a message from this channel to create a thread.");
+            else if ((archiveAfter == AutoArchiveDuration.ThreeDays && this.Guild.Features.Contains("THREE_DAY_THREAD_ARCHIVE")) || (archiveAfter == AutoArchiveDuration.Week && this.Guild.Features.Contains("SEVEN_DAY_THREAD_ARCHIVE")))
+                throw new ArgumentException("This archive duration requires the guild to be boosted or have these archive durations enabled."); //are guild features always cached?
+
+            return this.Discord.ApiClient.CreateThreadFromMessageAsync(this.Id, message.Id, name, archiveAfter, reason);
+        }
+
+        /// <summary>
+        /// Creates a new thread within this channel..
```
Comment (IDoEverything) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:945]: double full stop

**Review by IDoEverything — CHANGES_REQUESTED** [review]

**Comment by IDoEverything** [comment]
Also the missing endpoints

**Comment by IDoEverything** [comment]
https://github.com/discord/discord-api-docs/commit/7ab5731e97262bcfdeb9278f8add0d555b98478d

File: DSharpPlus/Entities/Channel/Thread/DiscordThreadChannel.cs (hunk 15 above)
Comment (DWaffles) [reply DSharpPlus/Entities/Channel/Thread/DiscordThreadChannel.cs:65]: Yes, I was under the impression I changed it, will be changing it to just the base type.

File: DSharpPlus/Clients/DiscordClient.Events.cs (hunk 12 above)
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Events.cs]: Corrected.

File: DSharpPlus/Clients/DiscordShardedClient.Events.cs (hunk 13 above)
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordShardedClient.Events.cs]: Corrected.

File: DSharpPlus/Entities/Channel/DiscordChannel.cs (hunk 14 above)
Comment (DWaffles) [reply DSharpPlus/Entities/Channel/DiscordChannel.cs]: Corrected.

File: DSharpPlus/Entities/Channel/Thread/DiscordThreadChannelMetadata.cs (hunk 16 above)
Comment (DWaffles) [reply DSharpPlus/Entities/Channel/Thread/DiscordThreadChannelMetadata.cs]: Fixed.

**Review by IDoEverything — APPROVED** [review]
From my minimal testing of the events and endpoints, this seems to work fine. I won't be merging this until other people review it just to make sure there isn't anything I missed.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 20]
```diff
@@ -1764,6 +1821,175 @@ internal async Task OnVoiceServerUpdateEventAsync(string endpoint, string token,
...
+                    ParentId = cthread.ParentId,
+                    Id = cthread.Id,
+                    Name = cthread.Name,
+                    Type = cthread.Type,
+                    LastMessageId = cthread.LastMessageId,
+                    MessageCount = cthread.MessageCount,
+                    MemberCount = cthread.MemberCount,
+                    ThreadMetadata = cthread.ThreadMetadata,
+                    CurrentMember = cthread.CurrentMember,
+                    //threadMembers,
```
Comment (IDoEverything) [inline DSharpPlus/Clients/DiscordClient.Dispatch.cs:1865]: is this missing properties?

File: DSharpPlus/Clients/DiscordClient.cs [hunk 21]
```diff
@@ -799,6 +817,7 @@ private void UpdateMessage(DiscordMessage message, TransportUser author, Discord
                 message.Author = this.UpdateUser(usr, guild?.Id, guild, member);
             }
 
+            //var channel = this.InternalGetCachedThread(message.ChannelId) ?? this.InternalGetCachedChannel(message.ChannelId);
```
Comment (IDoEverything) [inline DSharpPlus/Clients/DiscordClient.cs:820]: what's this for?

Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.cs]: This was to pull from the thread cache for when a message is sent. It works either way (which is why the change was commented after I made it), but the intent seems clearer this way so I'll keep it.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs (hunk 20 above)
Comment (DWaffles) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs]: Misc. comment.
